### ISSUE TITLE
Provide a way to specify archives to use on a per-query basis

### ISSUE THREAD

**Issue by ikreymer** [oduwsdl/memgator#39]
I'd like to have the ability to only query a specific subset of archives, specified by id from the complete list set in the config. For example,

`memgator --archives ia,uknationalarchives 2014 http://example.com/`

would only include the archives idenitfied by `ia` and `uknationalarchives` labels.

This may be a bit more tricky to specify the HTTP interface since query is part of the url, but I'd be happy with a custom header or any way that is easiest.

The use case here would be to allow users to dynamically select which archives to include in Netcapsule

**Comment by ibnesayeed** [comment]
Last week @machawk1 was showing his interest on a similar feature where clients can configure what archives they want to query instead of the server deciding what to include. We were thinking about possible API to accommodate this additional information in each request. However, ID based filtering from the existing superset of archives is not what he was asking though, he was more interested in a way to specify a custom archives.json file from the client in each request. These IDs were intended for logging purpose and not essential for the system to function so they are not really exposed to clients. Additionally, an ad-hoc token/keyword based local identifier is not something that web scientists will be happy with.

I think a way to pass custom archive.json file from the client will take care of this use case as well while allowing the aggregator to aggregate any custom source that was otherwise not known to it. In case of Netcapsule, the interface that allows user to select what archives to include from can easily generate a custom archive.json file based on the user selection from an existing list or additions to the list.

One way to accommodate it in the API is to have a custom request header that passes a reference to a publicly accessible custom archive.json file or a `data` URI that has a base64 encoded JSON data in it. Another way is to use cookies for this purpose instead of a custom request header. Please bring in any other ideas.

**Comment by ikreymer** [comment]
Well, ideally, the archives could be specified per session, rather than per request even, as in my use case, the same archives would be used to load the embeds, etc.. once the browsing session has been created. I could see sending a custom `archives.json` file once per session and getting a cookie to refer to, but that just means you'd have to implement session management and track different archive lists.. Sending it with every request seems excessive, so hence the simple id list idea, which should be quite performant..
To put it another way, I'd be concerned that sending a full JSON with every request will have a negative performance impact.

**Comment by ikreymer** [comment]
Maybe I'm being overly concerned, but trying to make this as performant as possible, of course :)

**Comment by ibnesayeed** [comment]
Session based archive list customization was one of the two solutions I proposed to @machawk1 when he requested for this feature. Additionally, session based customization is still possible without making the session management a first class member of MemGator itself which I would like to keep outside of the scope of MemGator as much as possible and off-load it to a proxy or wrapper application that is using it. One way to implement it without having too much complexity is to establish a cookie based session and create a separate MemGator server instance for that particular session if not exists already. Additionally, have a way to kill the MemGator instance if it has not received any traffic for `X` seconds or in other words when the session expires. Another approach would be to setup an external dynamic archive.json file service that responds with a different dataset for different requests and MemGator always query that service for each request.

ID based filtering has more cons than pros:

* IDs are internal, not exposed publicly, and have local short names instead of URNs
* Assigning URNs/URIs to each archive as their identifier will increase the size of data
* The server has to expose the list of archive IDs in a header or on a special endpoint that will increase the number of round trips
* Limit the possibility of aggregating arbitrary archives and only allow filtering from the set of archives that are known to the server
* Having a URL pointing to a custom list will be shorter than having comma separated list of IDs
* If the JSON is sent as base64 encoded data URI it will not be very network intensive (especially when the service is running locally) when compared to the generally large response size
* The system can function with as little data in the JSON file as just the TimeMap, TimeGate, or both URIs (as needed) in each archive object. The name, ID, and many other attributes can be optional

**Comment by ikreymer** [comment]
I mean, I could spin up a new memgator container for every emulated browser container, but that seems excessive and unnecessary, when the most that is happening is selecting from a subset of archives?
Unless there is a more complicated use case than selecting from a known subset, the id list seems like a logic solution. I think the ID/url selection is the best solution for this use case...

**Comment by ibnesayeed** [comment]
MemGator service is quite lightweight hence a separate container for each session will not cause any remarkable overhead. In my opinion the tool should provide a more generic solution that caters to more generalized use cases instead of very specific cases.

Also, think about it, how often a client would try to limit the number of archives? One practical approach is to create MemGator instances based on the customizations instead of individual sessions. Let me elaborate it, say, many clients are using the default configuration then they will utilize a single instance, but one client customized the selection on the client side then generate a hash of the subset of archives in a sorted order and pass that as a cookie which your service can detect and create a MemGator instance for that combination. Any other client if uses the same combination will be served by the same MemGator instance which will be killed automatically if not used for a while. Instantiating MemGator instances is quite fast and lightweight.

> Unless there is a more complicated use case than selecting from a known subset...

If a central service is running controlled by someone else then the client would have no idea about the IDs and would prefer to provide a custom archive.json file if not happy with the unknown default set of archives aggregated. In contrast, if a local instance is running then network latency is not an issue and passing a serialized JSON in the request header will cause no overhead (roughly 2-3x of IDs if the IDs are URIs) while still being open for more general use cases.

**Comment by ibnesayeed** [comment]
I am not trying to ignore the use case or discouraging the possibility of the proposal, instead I am just trying to propose a generalized solution that solves not only this use case, but handles some other cases as well. Having so many niche cases and individual implementations will cause a mess to maintain it in long term and will cause serious interoperability issues with other aggregator implementations.

**Comment by ikreymer** [comment]
Creating custom instances, and tracking which combinations have been used before, and which haven't, is a whole lot of extra bookkeeping, either in memgator or separate application, for something that should be a very simple filter..

But, taking a step back, perhaps it'd be good to define what an aggregator actually is.. To me, an aggregator is an service/abstraction that knows about 1 to N defined archives, and can return aggregate information from 1 to N archives. To that affect, it would be great if it could also 'advertise' which archives it knows about. I think that could be a good step to make aggregation more distributed.

With this conception, if a client has some other list of archives they want to use, the client should instead run its own aggregator service. Sending a list of archives to a remote aggregator doesn't really make sense, imo. In fact, it may be interesting to have aggregators talk to each other so that they answer requests about which archives they know about, a start in some sort of aggregator 'discovery' system..

Based on this conception, returning a list of which archives an aggregator knows about and allow users to filter which ones they want to get results from makes perfect sense.

**Comment by ikreymer** [comment]
I think having an aggregator be 'fixed' over a set of archives has other practical considerations.. for example, the aggregator may be optimized for access (reusable connections, etc..) to these archives, or perform additional internal cacheing.. Submitting an arbitrary list of archives.json with each request may necessitate breaking any such optimizations.. Filtering may to some extent as well, but less so, since that can be done after the fact..

**Comment by ibnesayeed** [comment]
MemGator by design does not take care the responsibility of the archives it's going to aggregate from. No archives are hardcoded in it, they list of archives is always an external information. This aggregator has the philosophy of being a utility that consolidates the responses from various sources that speak Memento protocol. It does not have a global knowledge of those sources, they or either supplied as a local or remote file or via a dynamic service. With that philosophy in mind, it shouldn't care if the list of archives is provided at the time of instance creation, set for a session, periodically updated internal session via an administrative endpoint, or piggybacked in each request, as long as the given list consists of Memento services.

**Comment by ibnesayeed** [comment]
> To that affect, it would be great if it could also 'advertise' which archives it knows about.

That is certainly something that can be exposed at an endpoint, but I would think about it as a dynamic response that may show different results based on the query. I am thinking about the Archive Profiling based probability ranking service as well as client requested archive list in this perspective. Here are a few examples that will illustrate what I mean.

```
# Shows all the archives known by default when the server instance was run (or the global session was modified latter).
GET /archives

# Shows top K (already configurable by flags) archives that will be queried based on the requested URI-R and the Archive Profile Probability Ranking service.
GET /archives/{URI-R}

# Shows archives that will be used to serve the request based on the data passed from the client.
GET /archives [HEADER Memento-Aggregator-Sources: location/of/custom/archive/list]
```

**Comment by ibnesayeed** [comment]
> With this conception, if a client has some other list of archives they want to use, the client should instead run its own aggregator service. Sending a list of archives to a remote aggregator doesn't really make sense, imo.

There are cases where clients cannot run an MemGator instance. Think about Mink, the chrome extension. It has to point to some hosted service of aggregator. The tool might provide a way to select what sources it aggregates from. The interface will probably query the aggregator for the list of aggregated archives and give the user an option to filter from them (so far it is same as your current request and can be implemented using your proposed filter approach). However, the user might also have a local archive or an instance in of WebRecorder somewhere which needs to be aggregated for the user without telling rest of the world about it. This will only be possible if the client can provide a custom list of archives.

**Comment by ibnesayeed** [comment]
> In fact, it may be interesting to have aggregators talk to each other so that they answer requests about which archives they know about, a start in some sort of aggregator 'discovery' system.

I love the idea of such a knowledge base and peer-to-peer discovery mechanism, but there are many cases where one would not be interested in exposing their collection publicly such as private or personal archives, restricted or dark archives. However, I don't feel like adding this responsibility into an aggregator. Instead, I can think of a separate service that works as an [archive registry](http://labs.mementoweb.org/aggregator_config/archivelist.xml).

On that note, if we setup a separate archive registry which allows dynamic filtering then it will solve your proposal along the same lines you want it to solve. here is an example to illustrate that.

Say, we have an archive registry service that exposes the list of known archives in various formats such as XML, JSON and whatnot. It also allows means to add new sources. It allows various ways to query/filter based on various attributes of individual archives:

```
# All archives
GET http://example.com/archives/json/

# Only Memento compliant archives
GET http://example.com/archives/json/?memento=true

# Only archives selected by IDs
GET http://example.com/archives/json/?ids=ia,ukwa,proni

# All archives except by IDs
GET http://example.com/archives/json/?!ids=pt,is
```

Now, any of these URLs can be used as the source (requested by the client) and the MemGator service will fetch that list to fulfill the request.

```
GET /timemap/json/{URI-R}
Memento-Aggregator-Sources: http://example.com/archives/json/?ids=ia,ukwa,proni
```

This approach does everything you want and the way you want while still keeping the archive registry responsibility out of the MemGator scope. This approach will also nicely accommodate the archive profile based dynamic archive registry without any change in the tool.

**Comment by ibnesayeed** [comment]
> I think having an aggregator be 'fixed' over a set of archives has other practical considerations.. for example, the aggregator may be optimized for access (reusable connections, etc..) to these archives, or perform additional internal cacheing.. Submitting an arbitrary list of archives.json with each request may necessitate breaking any such optimizations.. Filtering may to some extent as well, but less so, since that can be done after the fact..

Such performance tweaks are important for cases where a single MemGator server instance is handling millions of hits every few minutes from various sources. Additionally, persistent connection is something that falls into the category of quality of service and supporting infrastructure at the other end which is difficult to keep track of and stay up to date. Also, persistent connection are helpful when relatively small response data is requested very frequently which may not be true in this case where the response data size is often large enough to make the benefit of persistent connection insignificant. Additionally, this can still be achieved by persisting connections to some known good archives and service every one else without a persistent connection. Each archive is independent from each other and there is nothing that forces us to have the persistent connection in a complete pool or none at all. There is no guarantee that all archives will support persistent connection.

**Comment by ibnesayeed** [comment]
On the matter of performance optimization, I remember various pieces of information available in LANL's archive registry that tells the sort-order, pagination status and similar things about each archive that can be utilized for certain optimizations, but I try not to use them as they are subject to change silently and someone has to keep them up-to-date. However, if the tool works on minimum necessities mandated by the Memento protocol then it will be more inclusive and will be able to handle individual instances without a lot of configuration knowledge about each archive (institutional or personal).

**Comment by ikreymer** [comment]
Thanks for all the responses. Yes, I think the archive registry abstraction is the component I'd like to see.. I think it makes sense to combine an archive registry + aggregator, but if you'd like to keep them separate, that should work too for now. I am not convinced that an aggregator w/o a registry would be all that useful, as users will need to get a list of archives from somewhere, and then naturally aggregate that data.. I suppose keeping things modular at this point won't hurt..

**Comment by ibnesayeed** [comment]
> I am not convinced that an aggregator w/o a registry would be all that useful, as users will need to get a list of archives from somewhere, and then naturally aggregate that data.

This is true that the aggregator without a registry is useless, but unfortunately there is no reliable registry service already existing which has a standard format of response with the promise of its persistence without breaking changes for long in future. As a result tying a binary tool (not a central web application) to something temporary will be dangerous and may cause many installed instances of the tool to stop functioning when the dependency service is gone. This is the reason why I am shipping an example minimal archives.json file that has only Memento compliant archives in it. And to make it work out of the box I have a hosted static copy of that file which is referred by default. However, that static file is under my personal homepage of the department domain that needs to be moved to somewhere else before the 1.0 release (#35). If such a persistent standard registry comes into existence, we can change the default to that in future releases.

**Comment by galsondor** [comment]
This is a long-standing optimization I have needed.  Here is the basic driver.  Several otherwise useful archives mangle embedded URIs irreversibly (e.g., WebCite).  This makes them unsuitable for my research and I exclude them.  Even though they can clearly be filtered out client-side, allowing them to be excluded will save time and bandwidth.

Of course a implementation mechanism must be decided upon.  I think it must be a header to avoid possible confusion with the URI-R.  (*X-Memgator-Archive-Filter* anyone?)

**Comment by ibnesayeed** [comment]
@galsondor, MemGator makes this task of selecting what archives you want to aggregate from very easy, but you will have to provide a custom archives list file as a flag when running the CLI or Server otherwise it will default to [this archives.json file](http://oduwsdl.github.io/memgator/archives.json). And in case of the server, it will persist for the uptime of the server.

This request however is about making it possible for the clients to specify a custom archives list in any request and a running server obey that. It will certainly be done via some custom request header, thanks for tossing a name.

**Comment by machawk1** [comment]
I have put some effort into this in the `https://github.com/machawk1/memgator` fork. The method of archival specification needs a little more consideration but results from my local `pywb` are included in a TimeMap from a (forked) MemGator instance when curl'd with the appropriate header. It's hacky but a First Step in accomplishing this, for which more work will soon but put forth.
